modal_server_vllm.py

```diff
@@ -0,0 +1,424 @@
+        status_code=status_code.value,
+    )
+
+async def check_length(request, input_ids, model_config):
+    if hasattr(model_config.hf_config, "max_sequence_length"):
+        context_len = model_config.hf_config.max_sequence_length
+    elif hasattr(model_config.hf_config, "seq_length"):
+        context_len = model_config.hf_config.seq_length
+    elif hasattr(model_config.hf_config, "max_position_embeddings"):
+        context_len = model_config.hf_config.max_position_embeddings
+    elif hasattr(model_config.hf_config, "seq_length"):
+        context_len = model_config.hf_config.seq_length
+    else:
+        context_len = 4096
+
+    token_num = len(input_ids)
+
+    if token_num + request.max_tokens > context_len:
+        return create_error_response(
+            HTTPStatus.BAD_REQUEST,
+            f"This model's maximum context length is {context_len} tokens. "
+            f"However, you requested {request.max_tokens + token_num} tokens "
+            f"({token_num} in the messages, "
+            f"{request.max_tokens} in the completion). "
+            f"Please reduce the length of the messages or completion.",
+        )
+    else:
+        return None
+
+@app.cls(gpu=GPU_CONFIG,timeout=60 * 10,
+    container_idle_timeout=60 * 10,
+    allow_concurrent_inputs=10, image=image)
+class Model:
+    @modal.enter()
+    def start_engine(self):        
+        model, tokenizer,engine_model_config = get_model()
+        self.model = model
+        self.tokenizer = tokenizer
+        self.engine_model_config = engine_model_config
+
+    @modal.method()
+    async def generate(self, request:ChatCompletionRequest):
+        from vllm.sampling_params import SamplingParams
+        from functionary.inference import enforce_tool_choice, prepare_messages_for_inference
+        from functionary.inference_stream import generate_openai_format_from_stream_async
+        from vllm.utils import random_uuid
+        from functionary.prompt_template import (      
+            PredefinedFuncTypes,      
+            get_prompt_template_from_tokenizer,
+        )
+        from vllm.outputs import RequestOutput
+        from functionary.prompt_template.prompt_template_v2 import get_random_tool_call_id
+
+        
+
+        if request.tools:
+            tools = enforce_tool_choice(
+                tool_choice=request.tool_choice, tools=request.tools
+            )
+            tools_or_functions = [item.dict() for item in tools]
+        elif request.functions:
+            tools = None
+            tools_or_functions = [item.dict() for item in request.functions]
+        else:
+            tools = None
+            tools_or_functions = []
+        # Remove any code_interpreter tools remaining
+        if tools:
+            tools = [tool for tool in tools if tool.type != "code_interpreter"]
+            tools_or_functions = [
+                tool for tool in tools_or_functions if tool["type"] != "code_interpreter"
```

Comment:
This should be after `prepare_messages_for_inference` as the function involves reading for the presence of code_interpreter tool.